File: apps/windows/DepartmentRuntime.cs
namespace Biupiu.Desktop;

public sealed record DepartmentRuntimeTarget(string Route, string PackageName, string? Scope);
public sealed record DepartmentRuntimeState(DepartmentRuntimeTarget Target, bool CanStart, string Reason);

public static class DepartmentRuntime
{
    public static DepartmentRuntimeState Resolve(RuntimeSession session, string route)
    {
        if (session.Status != "active") return new(Target(route), false, "INACTIVE_ACCOUNT");
        var allowed = route is "RND_OS" || session.Entitlements.Contains(route);
        return new(Target(route), allowed, allowed ? "AUTHORIZED" : "NO_ENTITLEMENT");
    }

    private static DepartmentRuntimeTarget Target(string route) => route switch
    {
        "SMART_FARMING" => new(route, "@biupiu/smart-farming", "SMART_FARMING"),
        "SMART_METAL_WORKSHOP" => new(route, "@biupiu/smart-metallurgy", "SMART_METAL_WORKSHOP"),
        "RND_OS" => new(route, "@biupiu/rnd-os", null),
        _ => new(route, "UNRESOLVED", null)
    };
}

File: apps/android/app/src/main/java/com/biupiu/rndos/DepartmentRuntime.kt
package com.biupiu.rndos

data class DepartmentRuntimeTarget(
    val route: String,
    val packageName: String,
    val scope: String?
)

data class DepartmentRuntimeState(
    val target: DepartmentRuntimeTarget,
    val canStart: Boolean,
    val reason: String
)

object DepartmentRuntime {
    fun resolve(session: RuntimeSession, route: String): DepartmentRuntimeState {
        if (session.status != "active") {
            return DepartmentRuntimeState(target(route), false, "INACTIVE_ACCOUNT")
        }

        val allowed = route == "RND_OS" || session.entitlements.contains(route)
        return DepartmentRuntimeState(
            target(route),
            allowed,
            if (allowed) "AUTHORIZED" else "NO_ENTITLEMENT"
        )
    }

    private fun target(route: String): DepartmentRuntimeTarget = when (route) {
        "SMART_FARMING" -> DepartmentRuntimeTarget(route, "@biupiu/smart-farming", "SMART_FARMING")
        "SMART_METAL_WORKSHOP" -> DepartmentRuntimeTarget(route, "@biupiu/smart-metallurgy", "SMART_METAL_WORKSHOP")
        "RND_OS" -> DepartmentRuntimeTarget(route, "@biupiu/rnd-os", null)
        else -> DepartmentRuntimeTarget(route, "UNRESOLVED", null)
    }
}
